### DISCUSSION TITLE
Connection Pooling with Prisma: Only One Connection Being Pooled

### DISCUSSION THREAD

**Discussion by JesseRacicot** [prisma/prisma#21604] [Q&A]
### Question

Hello everyone,
I've been working on a Node.js application using Prisma and have encountered a perplexing connection pooling issue. Even before upgrading my Node version, I was trying to observe Prisma's connection pooling behavior by making multiple simultaneous requests to my database. My expectation was to see multiple connections being pooled, but the metrics consistently show only one connection. As an attempt to resolve the issue, I upgraded from Node v16.16.0 to v18.16.1, but unfortunately, the issue persists.

My database is remotely hosted on DigitalOcean with the following specifications:
- 32 GB RAM / 8vCPU / 160 GB Disk
- Primary only, located in NYC1
- PostgreSQL version 15

Interestingly, when I inspect the "Connections" under "Insights" on DigitalOcean, I observe that my DB has had multiple connections open via various API requests. This leads me to believe that this might be an issue on the Prisma client side.

Here's what I've tried:
- Created a test script to observe the number of open connections at any given time while using prisma.$metrics.json() to observe the number of open connections. 
- I did this by making several hundred concurrent requests with my prisma client.
- I Introduced an artificial delay of 5 second (also tried increasing it) to ensure that connections stay open longer.
- Checked my Prisma configuration to ensure that the maximum pool size is set to a number greater than 1.

My connection string (with sensitive information redacted) is:
```
DB_CONSTR=postgresql://[USERNAME]:[PASSWORD]@[HOST]:[PORT]/[DB_NAME]?pgbouncer=true&connection_limit=10
```
And the configuration portion for my schema.prisma file:
```prisma
generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "windows", "debian-openssl-3.0.x"]
  engineType    = "binary"
  previewFeatures = ["metrics"]
}
datasource db {
  provider = "postgresql"
  url      = env("DB_CONSTR")
}
```
Despite this (along with the node upgrade) I'm still observing only one connection being pooled at a time.
Is this the expected behavior? Or might there be something I'm overlooking? Appreciate any insights or suggestions.

Thanks in advance!

**Comment by janpio** [comment]
What value exactly are you watching for "open connections"? Prisma Metrics are a preview feature that can generally have bug - but afaik the display of the currently open connections should work fine.

**Comment by JesseRacicot** [comment]
We noticed there was a giant performance decrease on our DB (hosted on DigitalOcean) something like two weeks ago with many dropped connections due to timeouts.  Also, within the insights, the number of open connections at any given moment never exceeds 34 (although our DB pool could support 300 connections).  Here is a snippet of the code I used to count that the number of open connections at any moment.

```
const testPooling = async () => {

    const invId = 123
    const allIds = repeat(invId, 50)

    const getOne = async (id) => {
        const metrics = (await prisma.$metrics.json())
        console.log(metrics)
        return prisma.invoice.findFirst({ where: { id }})
    }

    console.log('FETCHING')

    const allProms = await Promise.all(map(getOne, allIds))

    console.log('ALL PROMS', allProms.length)
}

testPooling()
```

This should just fetch the same invoice 50 times, although concurrently.  I would imagine that, in order to support the concurrent requests, the number of open connections in the DB would be greater than 1.  Yet, that console.log(metrics), I see that number of open connections is always equal to 1.

Note that I have played with the number of requests and the latency (i.e. placing a delay before each request) and still obtain the same outcome.

Thanks,
Jesse Racicot
________________________________

**Reply by janpio** [reply]
> Also, within the insights, the number of open connections at any given moment never exceeds 34 (although our DB pool could support 300 connections). 

How many connecions did you configure as the connection_limit of Prisma?

**Reply by JesseRacicot** [reply]
From my connection string, you could see that I have configured the limit to be 10 (an arbitrary choice above 1 to test the pooling)

DB_CONSTR=postgresql://[USERNAME]:[PASSWORD]@[HOST]:[PORT]/[DB_NAME]?pgbouncer=true&connection_limit=10

I could see often when a connection drops, the prisma client config is logged and it states that the connection limit is set to 10.

Everything seemingly points to a connection limit that is set to 10 although the performance/behaviour (along with the metrics insights) suggest that only one connection can remain open at a time.

**Reply by janpio** [reply]
If you are using DO's connection pooling feature, Prisma does not control how many connections are opened to your _database_. The pooler of Digital Ocean does, and Prisma does not have any influence on that. The pooler itself usually surfaces a few thousand possible connections to the clients talking to it.

`connection_limit=10` configures how many connections Prisma should manage to whatever it is talking to. If that is the DO pooler, then it will open up to these many connections if they are needed and will keep them open and use when queries are executed.

Can you share the error messages or metrics output you are getting _from Prisma_ that make you think Prisma's connection pool is not working?

**Reply by JesseRacicot** [reply]
As I mentioned above in the sameple, the metrics that I logged show that only one connection is open at a time (i.e. prisma_pool_connections_open = 1 at all times)

I think a better idea would be if you recommend a way in which I could test/confirm that the prisma connection pooling is working as intended, e.g., what metrics I should log, what values to expect, etc 

P.S.  Prior to even running these tests, our team noticed a huge performance dropoff with many logs that stated Prisma was routinely dropping connections

**Reply by janpio** [reply]
Then remove the external connection pooler from the equation, and you can directly monitor the connections actually open on the database. That is an external confirmation of what Prisma's connection pool is doing. If you actually create load and traffic that requires Prisma to open multiple connections to the database, up to your connection limit, it will do that.

**Reply by JesseRacicot** [reply]
I have tried without connection pooling on DO and it is the same results.  On a particular route we make many concurrent prisma queries.  I logged the metrics for each of these queries.  Every log of the metrics object looked as follows;

` counters: [
    {
      key: 'prisma_client_queries_total',
      labels: {},
      value: 0,
      description: 'Total number of Prisma Client queries executed'
    },
    {
      key: 'prisma_datasource_queries_total',
      labels: {},
      value: 0,
      description: 'Total number of Datasource Queries executed'
    },
    {
      key: 'prisma_pool_connections_open',
      labels: {},
      value: 1,
      description: 'Number of currently open Pool Connections'
    }
  ],
  gauges: [
    {
      key: 'prisma_client_queries_active',
      labels: {},
      value: 0,
      description: 'Number of currently active Prisma Client queries'
    },
    {
      key: 'prisma_client_queries_wait',
      labels: {},
      value: 0,
      description: 'Number of queries currently waiting for a connection'
    },
    {
      key: 'prisma_pool_connections_busy',
      labels: {},
      value: 0,
      description: 'Number of currently busy Pool Connections (executing a database query)'
    },
    {
      key: 'prisma_pool_connections_idle',
      labels: {},
      value: 10,
      description: 'Number of currently unused Pool Connections (waiting for the next pool query to run)'
    },
    {
      key: 'prisma_pool_connections_opened_total',
      labels: {},
      value: 1,
      description: 'Total number of Pool Connections opened'
    }
  ],
  histograms: [
    {
      key: 'prisma_client_queries_wait_histogram_ms',
      labels: {},
      value: [Object],
      description: 'Histogram of the wait time of all queries in ms'
    }
  ]
}`

From my knowledge, it seems as though all of these metrics suggest that prisma is not pooling connections, e.g., synchronously serving these queries.  Perhaps I shouldn't be looking at these metrics to confirm/verify the pooling is working correctly yet I would then need to know how exactly I could test the connection pooling.  

Thanks

**Reply by janpio** [reply]
As I wrote, look at the number of open connections at the database. That is the only real proof that multiple connections are open. Metrics or anything else that Prisma tells you could in theory be wrong or misleading.

Besides that, the metrics object that you posted looks like the Prisma Client instance you are requesting these from has not processed any queries. Was this just a random example with numbers reset, or is that the _real_ result you are getting? Are you sure you are asking the same Prisma Client instance for metrics that is processing your queries?

**Reply by JesseRacicot** [reply]
I mentioned originally that the DB was handling more than one connection, but (likely) only because it was serving different clients (via seperate API requests).  But when it comes to handling one client, it is only opening one connection.

For example, on a particular API route I can make the same prisma calls 50 times (concurrently) and see that only one connection will be opened.  But if i make 50 requests to the same API route I will see 50 DB connections.  This to me suggests that DO is handling the pooling correctly but my prisma client is not.

As for my prisma client instance, one instance is opened when the server starts running and every API route imports this instance to make necessary requests.

**Reply by janpio** [reply]
I quickly took the code snippet you shared before, made it work and put it and the resulting metrics output on GitHub:
https://github.com/janpio/21604 I used a connection string with `?connection_limit=10`. As expected, the script created 10 connections to run these 100 queries, and at the end 10 connections were open and idle when I got the last metrics.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
